[PTRun] PowerToys Run - plugin dependency loading

CoreyHayward opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This PR builds on the open but seemingly no longer being worked on PR #18437 and implements the dependency loading for PTRun plugins from https://learn.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support allowing PTRun plugins to reference and make use of NuGet package dependecies.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #18435 
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [x] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
This change implements the recommended approach to dependency loading as detailed in this tutorial -> https://learn.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support

In addition a new plugin.props was introduced which when included in the plugins .csproj prevents the referenced projects dependencies being copied into the plugin output which caused issues when loading as they were treated as different types to the main IPlugin due to being loaded separately. see [here](https://github.com/dotnet/msbuild/issues/4371#issuecomment-1524551759) for a similar issue and why this was necessary.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
I manually validated this by including a NuGet package (Newtonsoft.Json) into one of the existing plugins and used it to deserialize a string.

coreyh-tigereye committed:
Updated the Wox.Plugin to correctly load and resolve plugin dependecies

coreyh-tigereye committed:
Included new plugin.props in all plugins to enable dynamic dependecy loading

CoreyHayward commented:
@microsoft-github-policy-service agree

coreyh-tigereye committed:
Updated dev docs to include new plugin.props in plugins

htcfreek commented:
@CoreyHayward 
Does your PR need the changes from #18437 too or does it work standalone?

CoreyHayward commented:
@htcfreek Changes work completely standalone

htcfreek commented:
Are the dependencies to load listed in the program code or do we load various dlls form the same path as the plugin file? Asking for security concerns/reasons.

CoreyHayward commented:
@htcfreek This uses the AssemblyDependencyResolver to load the dependencies dynamically and scoped to the context of the individual plugins.

From the [Microsoft page](https://learn.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support#load-plugins)
```
The AssemblyDependencyResolver object is constructed with the path to a .NET class library. It resolves assemblies and native libraries to their relative paths based on the .deps.json file for the class library whose path was passed to the AssemblyDependencyResolver constructor. The custom AssemblyLoadContext enables plugins to have their own dependencies, and the AssemblyDependencyResolver makes it easy to correctly load the dependencies.
```

jaimecbernardo commented:
Hi @CoreyHayward,
Thanks for the contribution!
I gave it a test and it worked fine mostly.
There's an issue with double loading if you enable the Terminal plugin, which I guess just needs this fix that was also applied in Flow https://github.com/Flow-Launcher/Flow.Launcher/blob/7fb8cc5642db22945bc4a40459586eedab6c0ae4/Flow.Launcher.Core/Plugin/PluginAssemblyLoader.cs#L33 , as @davidegiacometti  pointed to me while we were discussing this PR

One of the things we've noticed is that installed space increased a lot because of the dlls that are now in the output folder. tje installed RunPlugins folder went from ~12MB in 0.76.2 (including the localization files) to ~165 MB when I build the local installer (that doesn't contain the localization files). Even if some of those DLLs aren't needed, picking files manually would mean we needed to go back to picking files manually for the installer, reducing installer maintenance again. Is there a way to make the new logic optional for third-party plugins to opt in into?  Perhaps some plugin.json field to opt into this?

I'm assuming that's the real reason for this PR, right?

CoreyHayward commented:
@jaimecbernardo Thanks a lot for taking a look at this! 
Ill have a look into the bug with double loading, thanks for the fix reference!

Ill also take a look into making this an opt in thing, rather than a required change for every plugin. understand that it actually isn't needed for a lot of the more basic plugins that don't require any 3rd party dependencies and adds a fair amount of bloat to output of all other plugins too.

CoreyHayward committed:
Fixed double dependecy loading bug

CoreyHayward committed:
- Updated to only use dynamic loading when explicitly set by the plugin.
- Removed no longer required props from default plugins which do not need dynamic loading.
- Updated dev-docs to align with latest changes

CoreyHayward committed:
Removed line spacing changes in plugins csproj

CoreyHayward committed:
fixed spelling

CoreyHayward committed:
csproj cleanup

CoreyHayward committed:
removed unnecessary null checking

stefansjfw reviewed (approved):
Looks good to me. I gave it a test with using Newtonsoft.Json package in Calculator plugin. it works well. Size is 4MB more than current main which is expected because of the dependencies in Calculator directory.

Great contribution!
